### ISSUE TITLE
Behaviour change in datetime parsing for timestamps exactly at midnight

### ISSUE THREAD

**Issue by darylweir** [pydantic/pydantic#7039]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Accidentally uncovered a bug in our FastAPI application related to the way Pydantic parses dates/datetimes from path parameters.

In particular, we have an API that accepts either dates or datetimes as a path parameter, and behaves slightly differently depending on which it gets. (Arguably this is bad design, and indeed we've deprecated the API, but getting people to move to the new API is non-trivial... 😅). The underlying model has a variable of type `date | datetime`. 

We recently upgrade to Pydantic v2, and noticed that the parsing behaviour has changed for timestamps which fall exactly at midnight. In v1, such a string would parse into a datetime, but now in v2 it parses to a date.

This is not a major issue, since it only crops up when using a union of `date` and `datetime` as a field type, but it is a breaking change so thought I'd draw attention to it.

### Example Code

```Python
from datetime import date, datetime

from pydantic import BaseModel

class A(BaseModel):
   a: date | datetime

thing = A(a="2020-01-01T00:00:00")

# In v1, thing.a is datetime(2020, 1, 1, 0, 0)
# In v2, it's date(2020, 1, 1)
```

### Python, Pydantic & OS Version

```Text
             pydantic version: 2.1.1
        pydantic-core version: 2.4.0
          pydantic-core build: profile=release pgo=false mimalloc=true
                 install path: /Users/xxx/Library/Caches/pypoetry/virtualenvs/xxx/lib/python3.11/site-packages/pydantic
               python version: 3.11.1 (main, Mar  2 2023, 16:44:37) [Clang 11.0.3 (clang-1103.0.32.59)]
                     platform: macOS-10.15.7-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @davidhewitt

**Comment by MarkusSintonen** [comment]
@darylweir would enabling `strict`-mode help here? https://docs.pydantic.dev/latest/usage/strict_mode/#strict-mode-with-field

**Comment by MarkusSintonen** [comment]
(This is where the datetime->date coercion happens https://github.com/pydantic/pydantic-core/blob/1bd36f1ff6bc0a73085ce807e3fc139d218cf978/src/validators/date.rs#L123)

**Comment by darylweir** [comment]
@MarkusSintonen that doesn't seem to work, unfortunately.

```
class A(BaseModel):
   a: date | datetime = Field(strict=True)

A(a="2020-01-01T00:00:00")  # A(a=datetime.date(2020, 1, 1))
```

**Comment by MarkusSintonen** [comment]
There might be two issues at play 🤔  So only this works:

```python
class A(BaseModel):
    a: Annotated[date, Field(strict=True)] | Annotated[datetime, Field(strict=True)]

assert A.model_validate_json(json.dumps({"a": "2023-01-01T00:00:00"})).a == datetime(2023, 1, 1)
assert A.model_validate_json(json.dumps({"a": "2023-01-01"})).a == date(2023, 1, 1)
```

But it wont work with `model_validate` because this doesnt include `bytes_as_date` handling: https://github.com/pydantic/pydantic-core/blob/3f7df809123b643f84b9b5e39c008d9300692462/src/input/input_python.rs#L536 Im not sure why that is the case that in strict mode it wont allow parseable dates (only the exact instance). 🤔 

Also another question is why the `strict=True` only applies to the "outer union type" and not the members of it...

**Comment by MarkusSintonen** [comment]
> I think the big picture question here is if the current behavior is correct or not. I'm actually inclined to say it's incorrect, it's discarding information. If there is a timestamp I don't think it should parse to a date even if the timestamp is at midnight.

Yeah I agree that the current coercion in lax mode feels wrong. Its also inconsistent with Python date

**Comment by MarkusSintonen** [comment]
> In general the intention with strict mode is to avoid coercions. For example strict mode for int does not accept strings.

Yeah I see so its eg to forbid this `MyModel(v="1")` in strict mode as it wouldnt even make sense from type checking perspective.

> I think the big picture question here is if the current behavior is correct or not. I'm actually inclined to say it's incorrect, it's discarding information. If there is a timestamp I don't think it should parse to a date even if the timestamp is at midnight.

Agreed, it feels strange that a ISO datetime with (midnight) time gets magically coerced into a date. Its also inconsistent with Python:
```
>       date.fromisoformat("2023-01-01T00:00:00")
E       ValueError: Invalid isoformat string: '2023-01-01T00:00:00'
``` 

Can we remove the magical coercion of midnight-datetimes before its too late? 😅

**Comment by MarkusSintonen** [comment]
> Its also inconsistent with Python

Furthermore its inconsistent with json-schema validators. It fails in json schema validators for `date`-props.

**Comment by izeigerman** [comment]
I'd like add that when attempting to serialize such model to json, the `date` is serialized even if the datetime object is not midnight:

Model:
```
class Test(BaseModel):
    ts: t.Union[datetime.date, datetime.datetime]
```
Output:
```
In [8]: Test(ts=datetime.datetime(2023, 8, 9, 1)).json()
Out[8]: '{"ts":"2023-08-09"}'
```

**Comment by davidhewitt** [comment]
> > Its also inconsistent with Python
> 
> Furthermore its inconsistent with json-schema validators. It fails in json schema validators for `date`-props.

Agreed this is not great, opened #7065

**Comment by davidhewitt** [comment]
Thanks @izeigerman that's a separate issue which should be resolved by https://github.com/pydantic/pydantic-core/pull/873
